lib/icinga/checkable-dependency.cpp

```diff
@@ -55,66 +214,53 @@ bool Checkable::IsReachable(DependencyType dt, Dependency::Ptr *failedDependency
 
 		if (host && host->GetState() != HostUp && host->GetStateType() == StateTypeHard) {
-			if (failedDependency)
-				*failedDependency = nullptr;
-
 			return false;
 		}
 	}
 
-	auto deps = GetDependencies();
-
-	std::unordered_map<std::string, Dependency::Ptr> violated; // key: redundancy group, value: nullptr if satisfied, violating dependency otherwise
-
-	for (const Dependency::Ptr& dep : deps) {
-		std::string redundancy_group = dep->GetRedundancyGroup();
-
-		if (!dep->IsAvailable(dt)) {
-			if (redundancy_group.empty()) {
+	for (auto& dependencyGroup : GetDependencyGroups()) {
+		if (!(dependencyGroup->GetState(dt, rstack + 1) & DependencyGroup::State::ReachableOK)) {
+			if (dependencyGroup->IsRedundancyGroup()) { // For non-redundant groups, this should already be logged.
 				Log(LogDebug, "Checkable")
-					<< "Non-redundant dependency '" << dep->GetName() << "' failed for checkable '" << GetName() << "': Marking as unreachable.";
-
-				if (failedDependency)
-					*failedDependency = dep;
-
-				return false;
+					<< "All dependencies in redundancy group '" << dependencyGroup->GetName() << "' have failed for checkable '"
+					<< GetName() << "': Marking as unreachable.";
 			}
-
-			// tentatively mark this dependency group as failed unless it is already marked;
-			//  so it either passed before (don't overwrite) or already failed (so don't care)
-			// note that std::unordered_map::insert() will not overwrite an existing entry
-			violated.insert(std::make_pair(redundancy_group, dep));
-		} else if (!redundancy_group.empty()) {
-			violated[redundancy_group] = nullptr;
+			return false;
 		}
 	}
 
-	auto violator = std::find_if(violated.begin(), violated.end(), [](auto& v) { return v.second != nullptr; });
-	if (violator != violated.end()) {
-		Log(LogDebug, "Checkable")
-			<< "All dependencies in redundancy group '" << violator->first << "' have failed for checkable '" << GetName() << "': Marking as unreachable.";
-
-		if (failedDependency)
-			*failedDependency = violator->second;
+	return true;
+}
 
+/**
+ * Checks whether the last check result of this Checkable affects its child dependencies.
+ *
+ * @return bool - Returns true if the Checkable affects its child dependencies, otherwise false.
+ */
+bool Checkable::AffectsChildren() const
+{
+	auto cr(GetLastCheckResult());
+	if (!cr || IsStateOK(cr->GetState()) || !IsReachable()) {
+		// If there is no check result, the state is OK, or the Checkable is not reachable, we can't
+		// safely determine whether the Checkable affects its child dependencies.
 		return false;
 	}
 
-	if (failedDependency)
-		*failedDependency = nullptr;
+	for (auto& dep: GetReverseDependencies()) {
+		if (!dep->IsAvailable(DependencyState)) {
+			// If one of the child dependency is not available, then it's definitely due to the
+			// current Checkable state, so we don't need to verify the remaining ones.
+			return true;
+		}
+	}
 
-	return true;
+	return false;
 }
```

Comment:
There should be some definition of what "affects children" actually means. Like one can certainly argue that if the checkable is unreachable, this propagates down to the children which in some way affects the children.